torchdata/nodes/samplers/multi_node_weighted_sampler.py

```diff
@@ -0,0 +1,224 @@
+        if set(self.dataset_names) != set(self.weights.keys()) or len(self.dataset_names) != len(self.weights):
+            raise ValueError(
+                f"Invalid {self.weights=}. For multi-dataset weighted sampling, keys of source_nodes and weights must be the same",
+            )
+
+        for weight in self.weights.values():
+            if not isinstance(weight, float) or weight <= 0:
+                raise ValueError(
+                    f"""Invalid {self.weights=}. For multi-dataset weighted sampling, weights must be a 1d sequence, non-negative, and non-zero.
+                    Weights are used to sample from source nodes. Zero weight means the source node will never be sampled from, and can cause
+                    unexpected behavior depending on the stop criteris. Weights are used as inputs to torch.multinomial, please refer to
+                    https://pytorch.org/docs/stable/generated/torch.multinomial.html on how to use weights for sampling."""
+                )
+
+    def reset(self, initial_state: Optional[Dict[str, Any]] = None):
+        super().reset(initial_state)
+
+        if initial_state is not None:
+            self._num_yielded = initial_state[self.NUM_YIELDED_KEY]
+            self._weighted_sampler = _WeightedSampler(
+                weights=self.weights,
+                seed=self.seed,
+                rank=self.rank,
+                world_size=self.world_size,
+                initial_state=initial_state[self.WEIGHTED_SAMPLER_STATE_KEY],
+            )
+            self._datasets_exhausted = initial_state[self.DATASETS_EXHAUSTED_KEY]
+            for k in self.dataset_names:
+                self.source_nodes[k].reset(initial_state[self.DATASET_NODE_STATES_KEY][k])
+        else:
+            # Force a fresh iterator from all source nodes
+            self._num_yielded = 0
+            self._weighted_sampler = _WeightedSampler(
+                weights=self.weights,
+                seed=self.seed,
+                rank=self.rank,
+                world_size=self.world_size,
+            )
+            self._datasets_exhausted = {key: False for key in self.weights.keys()}
+            for k in self.dataset_names:
+                self.source_nodes[k].reset()
+
+    def _check_for_stop_iteration(self) -> None:
+        if all(self._datasets_exhausted.values()):
+            # Raise StopIteration if all datasets are exhausted,
+            # this covers for both ALL_DATASETS_EXHAUSTED and CYCLE_UNTIL_ALL_DATASETS_EXHAUSTED
+            raise StopIteration()
+
+        # Raise StopIteration is StopCriteria is FIRST_DATASET_EXHAUSTED and
+        # the first dataset is exhausted. Doing this to correctly catch StopIteration
+        # when trying next(it) on already exhausted iterator
+        if self.stop_criteria == StopCriteria.FIRST_DATASET_EXHAUSTED and any(self._datasets_exhausted.values()):
+            raise StopIteration()
+
+        return
+
+    def next(self) -> T:
```

Comment:
@andrewkho @ramanishsingh Update this method to remove the recursive call, but made the entire next() run inside a while loop which breaks after the first iteration - unless there is a redundant key in which cases to goes to the top and starts over. I am using a while loop so that we can check for early stop iteration (also moved the stop iteration checks out in a separate function). LMK if this make sense - will be adding more doc strings in next commit. Thanks.

Reply:
This is looking good, will mostly be relying on tests to ensure this works as intended